casanovo/denovo/model.py

```diff
@@ -171,16 +178,613 @@ def forward(
+        is_beam_prec_fit = (batch * beam) * [False]
+
+        # Create cache for decoded beams
+        (
+            cache_scores,
+            cache_tokens,
+            cache_next_idx,
+            cache_pred_seq,
+            cache_pred_score,
+        ) = self._create_beamsearch_cache(scores, tokens)
+
+        # Create the first prediction
+        pred, _ = self.decoder(None, precursors, memories, mem_masks)
+        _, idx = torch.topk(pred[:, 0, :], beam, dim=1)
+        tokens[:, 0, :] = idx
+        scores[:, :1, :, :] = einops.repeat(pred, "B L V -> B L V S", S=beam)
+
+        # Make precursors and memories the right shape for decoding:
+        precursors = einops.repeat(precursors, "B L -> (B S) L", S=beam)
+        mem_masks = einops.repeat(mem_masks, "B L -> (B S) L", S=beam)
+        memories = einops.repeat(memories, "B L V -> (B S) L V", S=beam)
+
+        # The main decoding loop:
+        for idx in range(1, self.max_length + 1):
+            scores = einops.rearrange(scores, "B L V S -> (B S) L V")
+            tokens = einops.rearrange(tokens, "B L S -> (B S) L")
+            # Terminate beams exceeding precursor m/z tolerance
+            # and track all terminated beams
+            finished_beams_idx, tokens = self._terminate_finished_beams(
+                tokens, precursors, is_beam_prec_fit, idx
+            )
+            # Cache terminated beams, group and order by fitting precursor m/z
+            # and confidence score
+            self._cache_finished_beams(
+                finished_beams_idx,
+                cache_next_idx,
+                cache_pred_seq,
+                cache_pred_score,
+                cache_tokens,
+                cache_scores,
+                tokens,
+                scores,
+                is_beam_prec_fit,
+                idx,
+            )
+            # Set all beams back to False after caching is completed
+            is_beam_prec_fit = (batch * beam) * [False]
+
+            # Stop decoding when all current beams are terminated
+            decoded = (tokens == self.stop_token).any(axis=1)
+            if decoded.all():
+                # Return the peptide with highest confidence score, within
+                # precursor m/z tolerance if possible
+                output_tokens, output_scores = self._get_top_peptide(
+                    cache_pred_score, cache_tokens, cache_scores, batch
+                )
+                break
+
+            # Get scores
+            scores[~decoded, : idx + 1, :], _ = self.decoder(
+                tokens[~decoded, :idx],
+                precursors[~decoded, :],
+                memories[~decoded, :, :],
+                mem_masks[~decoded, :],
+            )
+
+            # Find top-k beams with highest scores and continue decoding those
+            scores, tokens = self._get_topk_beams(scores, tokens, batch, idx)
+
+        return self.softmax(output_scores), output_tokens
```

Comment:
**issue:** In the unrealistic situation that decoding of one or more beams didn't finish within `max_length` steps, `output_scores` and `output_tokens` will be undefined. Should we move the `return` up to within the `if decoded.all()` and appropriately handle the edge case when decoding didn't fully finish here?

Reply:
How should we handle case if decoding didn't fully finish?

Reply:
My suggestion: if at least one beam finished, select from the finished beams. If no beam finished at all, just return dummy values (zero). We can write a unit test for this behavior by setting `max_length` to some ridiculously small number to verify that our implementation is correct, although this will likely never occur in reality (or at least it shouldn't).

Reply:
Made the changes you requested and implemented a test controlling if a finished beam is return if there are any. However, couldn't find a straightforward way to test if dummy values returned when `max_length` very small. Should I initialize a fixed model with known output? If not feel free to resolve the conversation

Reply:
@bittremieux in case you haven't seen my earlier comment

Reply:
I added a test case.